SortingStep: deduce way to sort based on input stream sort description

devcrafter opened the pull request:
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Deduce way to sort based on input stream sort description. Skip sorting if input stream is already sorted

devcrafter committed:
Skip sorting step if input stream is globally sorted

devcrafter committed:
In case full sort was wrong choise during plan interpretation

devcrafter commented:
@Mergifyio update

devcrafter committed:
SortingStep: type of sorting is deduced based on input stream sorting description in during transformation

+ perf test

devcrafter committed:
Fix + SortMode::None as default value

devcrafter committed:
Fix: prefix size check

devcrafter committed:
Code polishing

devcrafter committed:
Try to choose sorting transform based on sort description with fallback

devcrafter committed:
Fix: forgot to return sorting type in constructors

devcrafter committed:
Fix: test for optimize read in window order

+ code polishing

devcrafter commented on src/Processors/QueryPlan/Optimizations/reuseStorageOrderingForWindowFunctions.cpp:
```diff
@@ -102,10 +102,7 @@ size_t tryReuseStorageOrderingForWindowFunctions(QueryPlan::Node * parent_node,
     if (order_info)
-    {
         read_from_merge_tree->setQueryInfoInputOrderInfo(order_info);
-        sorting->convertToFinishSorting(order_info->order_key_prefix_descr);
```
It shouldn't be necessary any more, - we should get sort description from input stream on sorting step. @novikd please could you have a look?

devcrafter committed:
Fix: self-review comments

devcrafter committed:
Fix: consider collation in column sort description comparison

devcrafter committed:
Merge branch 'master' into skipping_sorting_step

devcrafter committed:
Fix: collator comparison in sort description

devcrafter committed:
Use sort mode Port for reading in order

devcrafter committed:
Fix: 01655_plan_optimizations_optimize_read_in_window_order_long

basically I returned code. Both plans with Finish sorting, need to check
sorting prefix

devcrafter committed:
clean up

devcrafter committed:
Merge branch 'master' into skipping_sorting_step

devcrafter committed:
Revert: 01655_plan_optimizations_optimize_read_in_window_order.reference

devcrafter committed:
Introduce Auto mode for sorting step (replace others for now)

devcrafter committed:
If sorting type is specified then use it. Otherwise rely on sort description

devcrafter committed:
Disable finish sort with sorted chunks

devcrafter committed:
Merge branch 'master' into skipping_sorting_step

devcrafter committed:
Fix: SortMode::Chunk

devcrafter committed:
Self-review + revert test

devcrafter committed:
Perf test. Code polishing

devcrafter committed:
Merge branch 'master' into skipping_sorting_step

devcrafter committed:
Just to trigger CI

devcrafter committed:
Merge remote-tracking branch 'origin/master' into skipping_sorting_step

devcrafter committed:
Update perf tests (to test perf test)

devcrafter committed:
optimize_sorting_for_input_stream setting and perf tests

devcrafter committed:
Merge branch 'master' into skipping_sorting_step

devcrafter commented on tests/performance/sorting_step_sort_descr_adoption.xml:
```diff
@@ -0,0 +1,19 @@
+<test>
```
Probably rename file according to setting name

devcrafter commented:
@nickitat @KochetovNicolai would you mind to review?

devcrafter committed:
rename and fix perf test

nickitat commented:
some of the possible bad cases from the chat with Vladimir
https://pastila.nl/?009cde16/669061285f5478a836a4837d2b41ca00
added just to not be lost from the context

nickitat commented:
`distinct_in_order` perf test seem to be flaky

nickitat commented on src/Processors/QueryPlan/SortingStep.cpp:
```diff
@@ -111,129 +114,186 @@ void SortingStep::convertToFinishSorting(SortDescription prefix_description_)
-
-            pipeline.addTransform(std::move(transform));
+            LOG_DEBUG(getLogger(), "Almost FullSort");
+            fullSort(pipeline, result_description, limit, true);
```
imo, this way would be more readable:
``` cpp
const bool partial_sort_not_needed = ...;
fullSort(pipeline, result_description, limit, partial_sort_not_needed);
```
without if

nickitat commented on src/Processors/QueryPlan/SortingStep.cpp:
```diff
@@ -111,129 +114,186 @@ void SortingStep::convertToFinishSorting(SortDescription prefix_description_)
+
+void SortingStep::fullSort(QueryPipelineBuilder & pipeline, const SortDescription & result_sort_desc, const UInt64 limit_, const bool skip_partial_sort)
+{
+    if (!skip_partial_sort)
```
I'm not sure that it will not introduce regression in some cases. `PartialSortingTransform` can throw out significant portion of rows due to the `limit`. pls, check it carefully (also taking [this](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Transforms/PartialSortingTransform.h#L44) into account), let's add a perf test for such case.

devcrafter committed:
Merge remote-tracking branch 'origin/master' into skipping_sorting_step

devcrafter committed:
Fix option description

devcrafter committed:
More generic getInputOrderInfo() in query info

devcrafter committed:
Merge remote-tracking branch 'origin/master' into skipping_sorting_step

devcrafter committed:
ExpressionStep: simple check if expression changes sorting order

devcrafter committed:
Merge remote-tracking branch 'origin/master' into skipping_sorting_step

devcrafter committed:
Test: check sorting with alias which match sorted column name

devcrafter committed:
Fix style

devcrafter committed:
Analyze ActionsDAG in ExpressionStep

+ check if sorting order can be kept from prevous step
+ EXPLAIN PLAN header=1 contains Sort Mode and Sort Description for each step
+ some tests
+ fix some review comments

devcrafter committed:
EXPLAIN PLAN sortmode options

+ ActionsDAG fixes

devcrafter committed:
Merge branch 'master' into skipping_sorting_step

devcrafter committed:
ActionsDAG check is not dependent on sort description

KochetovNicolai commented on src/Interpreters/ActionsDAG.cpp:
```diff
@@ -1923,4 +1924,34 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
     return actions;
 }
 
+bool ActionsDAG::isSortingPreserved() const
```
This property is a little bit strange.
It may be reasonable to make it dependent on some inputs and return outputs.

It is not very often that function is globally monotonic.
Even if all the functions are globally monotonic, we can remove a sorting key column from result.

KochetovNicolai commented on src/Processors/QueryPlan/ExpressionStep.cpp:
```diff
@@ -5,35 +5,40 @@
 }
 
-ExpressionStep::ExpressionStep(const DataStream & input_stream_, ActionsDAGPtr actions_dag_)
+ExpressionStep::ExpressionStep(const DataStream & input_stream_, const ActionsDAGPtr & actions_dag_)
```
Why?

KochetovNicolai commented on src/Processors/QueryPlan/ExpressionStep.cpp:
```diff
@@ -5,35 +5,40 @@
 
+static Poco::Logger * getLogger()
+{
+    static Poco::Logger & logger = Poco::Logger::get("ExpressionStep");
```
I think it's not a good idea to store logger statically.

KochetovNicolai commented on src/Processors/QueryPlan/SortingStep.h:
```diff
@@ -49,25 +50,35 @@ class SortingStep : public ITransformingStep
+        UInt64 limit_,
+        bool skip_partial_sort = false);
+
     enum class Type
```
Agree. 
I think that Full was better then Auto
Also I think we would better remove type at all

devcrafter replied:
When `SortingStep` constructed with `MergingSorted` or `FinishSorting`, - we consider that a caller has more information what sorting to apply, so we'll not try to infer it. And `Auto` is a mode when we infer the way to sort based on sorting properties of input stream. I'll make a comment

KochetovNicolai commented on src/Interpreters/InterpreterExplainQuery.cpp:
```diff
@@ -174,7 +174,8 @@ struct QueryPlanSettings
             {"optimize", optimize},
-            {"json", json}
+            {"json", json},
+            {"sortmode", query_plan_options.sort_mode},
```
I think it may be more reasonable to add a common setting which shows all the known data stream properties.
Right now it can be sort description + sort mode, distinct columns

devcrafter committed:
Check if expression step violates sorting correctly

devcrafter committed:
Try to fix style errors

devcrafter committed:
Check if sorting order is preserved for FilterStep

devcrafter committed:
Merge remote-tracking branch 'origin/master' into skipping_sorting_step

devcrafter commented on src/Core/SortDescription.cpp:
```diff
@@ -42,6 +42,19 @@ void SortColumnDescription::explain(JSONBuilder::JSONMap & map) const
+bool SortDescription::hasPrefix(const SortDescription & prefix) const
+{
+    if (prefix.empty() || prefix.size() > size())
+        return false;
```
ok

devcrafter commented on src/Core/SortDescription.cpp:
```diff
@@ -42,6 +42,19 @@ void SortColumnDescription::explain(JSONBuilder::JSONMap & map) const
     map.add("With Fill", with_fill);
 }
 
+bool SortDescription::hasPrefix(const SortDescription & prefix) const
```
Don't see how it fits. Prefix and description itself don't always have the same number of elements

devcrafter committed:
Fix review comments

devcrafter committed:
EXPLAIN tests

devcrafter committed:
Merge remote-tracking branch 'origin/master' into skipping_sorting_step

devcrafter commented on src/Interpreters/ActionsDAG.cpp:
```diff
@@ -1926,4 +1927,21 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
+    const Block & output_header = updateHeader(header);
+    for (const auto & desc : sort_description)
+    {
+        if (!output_header.findByName(desc.column_name))
```
Yes, it's not covered yet

devcrafter committed:
Consider aliases when checking if sorting order is preserved by
expression

qoega commented:
02377_optimize_sorting_for_input_stream_explain - need to update reference

KochetovNicolai commented on src/Processors/QueryPlan/SortingStep.cpp:
```diff
@@ -57,14 +60,14 @@ SortingStep::SortingStep(
     const DataStream & input_stream_,
-    SortDescription prefix_description_,
-    SortDescription result_description_,
+    const SortDescription & prefix_description_,
+    const SortDescription & result_description_,
```
Why this change is needed? Passing by value + std::move will avoid potential copy.

KochetovNicolai replied:
For example, here, if we add std::move to `order_descr`
https://github.com/ClickHouse/ClickHouse/blob/fe6533f41db18e7bb75a29f660c0fb2f68ee98d7/src/Interpreters/InterpreterSelectQuery.cpp#L1438

devcrafter replied:
Agree. It was oversimplification

KochetovNicolai commented on src/Processors/QueryPlan/SortingStep.cpp:
```diff
@@ -75,12 +78,12 @@ SortingStep::SortingStep(
 SortingStep::SortingStep(
     const DataStream & input_stream,
-    SortDescription sort_description_,
+    const SortDescription & sort_description_,
```
.

KochetovNicolai commented on src/Processors/QueryPlan/SortingStep.cpp:
```diff
@@ -111,129 +114,174 @@ void SortingStep::convertToFinishSorting(SortDescription prefix_description_)
-                    limit);
-
-            pipeline.addTransform(std::move(transform));
+            type = Type::MergingSorted;
```
Right now we never call transformPipeline more than once, but there may be good to allow it for some optimizations.
Let's try not to change internal state here.
(Maybe I should have done transformPipeline const)

devcrafter committed:
Fix test, review comments

devcrafter commented on src/Processors/QueryPlan/QueryPlan.cpp:
```diff
@@ -325,8 +325,22 @@ static void explainStep(
+    {
+        if (step.hasOutputStream())
+        {
+            settings.out << prefix << "Sort Mode: " << step.getOutputStream().sort_mode;
```
imo, the sorting properties are logically one thing (scope + how sorted). Internal separation into different variables is implementation detail. So, I'd keep it at one line. Regarding the formatting of this line - we can change it.
Also, one line allows checking the sorting in `explain` test much easier, - there is no need to check multiple lines.

@KochetovNicolai @vdimir opinions?

KochetovNicolai replied:
Generally agree with Nikita.
Also I would not print anything if sort mode is None

devcrafter replied:
Discussed with @KochetovNicolai 
Agreed to the following formating (one line):
```
Sorting (<scope>): <sort description>
```

devcrafter replied:
Also, regarding sorting scope (mode), we agreed that current scopes `Port` and `Stream` are confusing.
The proposal is to rename them (as follow-up):
`Port` -> `Stream`
`Stream` -> `Full` (or `Global`)

@nickitat opinion?

KochetovNicolai commented on src/Interpreters/ActionsDAG.cpp:
```diff
@@ -1926,4 +1927,49 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
+        }
+
+        /// check that found colunm is not an alias
+        for (const Node * node : index)
+        {
+            if (node->type == ActionType::ALIAS && node->result_name == desc.column_name)
+                return false;
+        }
```
I will try to create a counter-example for this

KochetovNicolai reviewed (commented):
Generally GLTM, but I will need to check `ActionsDAG::isSortingPreserved`

KochetovNicolai commented:
Found.
```
create table tab (x UInt64) engine = MergeTree order by x + 1;
insert into tab select number from numbers(4);
select x, y from (select x + 1 as y, sipHash64(x) as x from (select x from tab order by x + 1)) order by x + 1;

SELECT
    x,
    y
FROM
(
    SELECT
        x + 1 AS y,
        sipHash64(x) AS x
    FROM
    (
        SELECT x
        FROM tab
        ORDER BY x + 1 ASC
    )
)
ORDER BY x + 1 ASC

Query id: a032735e-1364-4193-8156-6511b478dd67

┌────────────────────x─┬────────────────────y─┐
│ 16738165381834614119 │ 16738165381834614120 │
│  9224715256000962398 │  9224715256000962399 │
│ 13686418376000424449 │ 13686418376000424450 │
│ 14247234200463187066 │ 14247234200463187067 │
└──────────────────────┴──────────────────────┘
```

```
set optimize_sorting_for_input_stream=0;
select x, y from (select x + 1 as y, sipHash64(x) as x from (select x from tab order by x + 1)) order by x + 1;

SELECT
    x,
    y
FROM
(
    SELECT
        x + 1 AS y,
        sipHash64(x) AS x
    FROM
    (
        SELECT x
        FROM tab
        ORDER BY x + 1 ASC
    )
)
ORDER BY x + 1 ASC

Query id: d4611d8c-890c-4ce0-82d5-396ed61b5592

┌────────────────────x─┬────────────────────y─┐
│  9224715256000962398 │  9224715256000962399 │
│ 13686418376000424449 │ 13686418376000424450 │
│ 14247234200463187066 │ 14247234200463187067 │
│ 16738165381834614119 │ 16738165381834614120 │
└──────────────────────┴──────────────────────┘
```

devcrafter committed:
Avoid copying SortDescription during SortingStep construction

devcrafter committed:
Merge remote-tracking branch 'origin/master' into skipping_sorting_step

devcrafter committed:
Merge remote-tracking branch 'origin/master' into skipping_sorting_step

devcrafter committed:
Fix compilation errors after fixing merge conflicts

devcrafter committed:
More robust check if sorting order is changed by expression

+ tests

devcrafter commented:
@KochetovNicolai Please check the fix. It addresses the case you provided

devcrafter committed:
More simple way to check if sorting order is preserved

- there is a case where it's done wrong

devcrafter committed:
Fix of check if sorting order is preserved

devcrafter committed:
Merge remote-tracking branch 'origin/master' into skipping_sorting_step

devcrafter committed:
Test improvements + more tests

devcrafter commented:
@KochetovNicolai @nickitat Please could you review it once more? It should be ready. I've added comments regarding possible follow-ups

KochetovNicolai commented on tests/queries/0_stateless/02377_optimize_sorting_for_input_stream_explain.reference:
```diff
@@ -37,7 +37,7 @@ Sort Mode: Stream: x ASC, y ASC
 Sort Mode: Chunk: a ASC, b ASC
 -- PLAN: actions chain breaks sorting order: input(column a)->sipHash64(column a)->alias(sipHash64(column a), a)->plus(alias a, 1)
-Sort Mode: Stream: plus(a, 1) ASC
+Sort Mode: None
```
(1) I think it's ok. Would be perfect to propagate this information, but it's not a error if we can't.
(2) Hard to say why it is happening. Probably some logic which removes columns uses only names. For a query plan (and ActionsDAG) it is allowed to have duplicate columns (duplicates are resolved in-order).

devcrafter replied:
I'll try to address it later as follow-up tasks

nickitat replied:
btw, we should be careful with functions like `plus`: `18446744073709551614 < 18446744073709551615`, but if we apply `plus(*, 1)` it became false.

KochetovNicolai commented on src/Interpreters/ActionsDAG.cpp:
```diff
@@ -1941,4 +1942,102 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
+    const Block & output_header = updateHeader(input_header);
+    for (const auto & desc : sort_description)
+    {
+        /// header contains column with the same name
+        if (output_header.findByName(desc.column_name))
+        {
+            /// find the corresponding node in output
+            const auto * output_node = tryFindInOutputs(desc.column_name);
+            if (!output_node)
+            {
+                /// sorted column name in header but NOT in expression output -> no expression is applied to it -> sorting preserved
+                continue;
+            }
+        }
```
This code is ok.
I think it is the same as to check if there is an input with this name.

devcrafter committed:
Sorting properties in EXPLAIN PLAN

~ change formatting for sorting
~ rename sortmode option -> sorting

devcrafter committed:
Fix: don't set sort mode in ReadFromMergeTree if sort description empty

devcrafter committed:
Merge remote-tracking branch 'origin/master' into skipping_sorting_step

nickitat commented on src/Interpreters/ActionsDAG.cpp:
```diff
@@ -1941,4 +1942,102 @@ ActionsDAGPtr ActionsDAG::cloneActionsForFilterPushDown(
+
+                /// TODO: we support monotonicity check only for functions with one parameter but ...
+                ///       if one parameter is variable and other are constant then we can try to check monotonicity as well
+                const auto monotonicity = func->getMonotonicityForRange(*types.front(), field, field);
```
in all three cases: here, at l.1974 and l.1991 imo should be provided some comments, why only `front()` is considered.

devcrafter commented:
> "global" is a good name. "stream", I am afraid, will lead to confusion between DataStream and streams on pipeline level. imo "port" is ok. and "stream" also ok for me.

It looks to me as opposite. Regarding parallel processing, we're saying that we have several data streams processed in parallel - if each stream is sorted, then sorting scope (we call it sort mode) is _stream_. Port is pipeline level implementation detail.

devcrafter commented on src/Core/Settings.h:
```diff
@@ -626,6 +626,7 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     M(Bool, compatibility_ignore_auto_increment_in_create_table, false, "Ignore AUTO_INCREMENT keyword in column declaration if true, otherwise return error. It simplifies migration from MySQL", 0) \
     M(Bool, multiple_joins_try_to_keep_original_names, false, "Do not add aliases to top level expression list on multiple joins rewrite", 0) \
     M(Bool, optimize_distinct_in_order, true, "Enable DISTINCT optimization if some columns in DISTINCT form a prefix of sorting. For example, prefix of sorting key in merge tree or ORDER BY statement", 0) \
+    M(Bool, optimize_sorting_for_input_stream, true, "Optimize sorting by sorting properties of input stream", 0) \
```
Then probably `optimize_sorting_by_input_stream_sorting_properties` ?

devcrafter committed:
Fix review comments

devcrafter committed:
Rename setting to optimize_sorting_by_input_stream_properties

devcrafter committed:
Rename tests according to new setting name

devcrafter commented:
Stress test failure is known issue #39857, see https://s3.amazonaws.com/clickhouse-test-reports/38719/fdd699e8a377c25bf8ff9e1d2c14f760cc79a4ee/stress_test__undefined_/fatal_messages.txt

devcrafter commented:
> > "global" is a good name. "stream", I am afraid, will lead to confusion between DataStream and streams on pipeline level. imo "port" is ok. and "stream" also ok for me.
> 
> It looks to me as opposite. Regarding parallel processing, we're saying that we have several data streams processed in parallel - if each stream is sorted, then sorting scope (we call it sort mode) is _stream_. Port is pipeline level implementation detail.

See proposal for renaming #40305
